README.md
# jquery-kpi-plugin

- Install the plugin using `npm i jquery-kpi-plugin`
- Call the plugin:
	-  ``` $(#selector).kpi({ params });```
	- for the selector, choose any DOM element with id
	- params are optional: if nothing is passed, engagement will be set to false
	- ex. ```$(#selector).kpi({ engagement: true });```

- At the moment the plugin has two modes, which can be adjusted as params passed to the plugin when called
  - engagement: true or false
  	- if engagement is set to false, the popover will appear within 2 seconds of page load; or, if the user tries to leave the page, the modal will be triggered again(whichever happens first)
  	- if the modal triggers, a cookie will be set to true and will prevent further triggering
  	- the cookie is set to expire in three days
  	- cookies are destroyed on quitting the browser
  	- if engagement is set to true, the popover will appear not sooner than approximately 40s or 2/3 of the time required to finish reading the page, whichever is shorter
  	- page word count is assessed on page load using p nodes

- List of config params and their defaults:
    - engagement: false,
    - pageCounter: 0,
    - timeToRead: 0,
    - fadeDuration: 250,
    - fadeDelay: 0,
    - pageHits: 0,
    - regex: ''


- `engagement` is false by default
- `pageCounter` sets counter for events to be triggered while on page
- `timeToRead` sets time to read text on page
- `fadeDuration` sets param for jquery modal (specify in milliseconds)
- `fadeDelay` sets param for jquery modal (specify in seconds)
- `pageHits` sets number of page hits before modal is triggered
- `regex` empty string by default

- Dependencies:
  - jquery, js-cookie, jquery-modal


CHANGELOG.md
## 1.0.0 - 2018-11-12

- Initial release

## 1.0.1 - 2019-02-13

- Make pageCounter param dynamic
- Add timeToRead  config param separate from the modal counter to check time elapsed for reading  and make it dynamic
- Add pageHits config param


## 1.0.2 - 2019-03-07

- Add option for newsletter subscription
- Exclude a page


## 1.0.4 - 2019-03-07

- Fix typo

## 1.0.5 - 2019-03-07

- Move up condition that checks subscription

## 1.0.6 - 2019-03-18

- set cookies to expire in 365 for newsletter signup otherwise 3 days
- better check for internal link navigation

## 1.0.7 - 2019-03-18

- set cookies as an inner var

## 1.0.8 - 2019-05-17

- update jQuery






kpi.js
;( function( $, window, document, undefined ) {

    "use strict";
        // Create the defaults once
        var pluginName = "kpi",
            defaults = {
                engagement: false,
                isCookieSet: false,
                mouseLeave: false,
                pageCounter: 0,
                timeToRead: 0
            };

        function Plugin ( element, options ) {
            this.element = element;
            this.settings = $.extend( {}, defaults, options );
            this._defaults = defaults;
            this._name = pluginName;
            this.init();
        }

        $.extend( Plugin.prototype, {
            init: function() {
                this.checkCookies();
                this.isMouseLeave();
                this.timeOut();
                this.counter();
                this.checkWordCount();
            },
            checkCookies: function( ) {
                // check cookies
                if (Cookies.get('modalpopupcookie') === undefined) {
                    var cval = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
                    // set cookie expiration time in 3 days
                    Cookies.set('modalpopupcookie', cval, { expires: 3, path: '' });
                } else {
                    this.settings.isCookieSet = true;
                }
            },

            // set default timeout for engagement false
            timeOut: function()
            {
                var engagement  = this.settings.engagement;
                var id = this.element.id;

                if (this.settings.isCookieSet === false
                    && engagement === false){
                    window.setTimeout(function(){
                        $('#'+id).modal();
                    }, 2000);
                }
                this.settings.isCookieSet = true;
            },

            isMouseLeave: function()
            {
                var mouseLeave = this.settings.mouseLeave;
                var engagement = this.settings.engagement;
                var isCookieSet = this.settings.isCookieSet;

                var id = this.element.id;

                $('body').mouseleave(function(){
                    if (mouseLeave === false && engagement === false && isCookieSet === false) {
                        $('#'+id).modal();
                        mouseLeave = true;
                    }
                });
                    this.settings.mouseLeave = mouseLeave;
            },

            counter: function ()
            {
                var pageCounter = this.settings.pageCounter;
                var id = this.element.id;
                var isCookieSet = this.settings.isCookieSet;
                var timeToRead = this.settings.timeToRead;

                var runCounter = function(){
                    if (pageCounter == 10) {
                        clearInterval(this);
                    }
                    else {
                         checkConditions();
                    }
                };
                // engagement is true
                var checkConditions = function () {
                    if (pageCounter >= 7 && isCookieSet !== true){
                        $('#'+id).modal();
                    } else if(pageCounter*7 > 0.66*(timeToRead)){
                        ('#'+id).modal();
                    }
                };

                setInterval(runCounter, 7000);
                runCounter();
            },

            // get word count using p nodes
            checkWordCount: function()
            {
                var timeToRead = this.settings.timeToRead;
                var wordsPerMinute = 250;
                var nodes = document.querySelectorAll('p'),i;
                var gross = "";
                var wordCount = null;
                if(nodes.length > 0){
                    for (i = 0; i < nodes.length; i++) {
                        gross += nodes[i].innerText;
                    }
                    wordCount = gross.replace(/(<([^>]+)>)/ig,"").split(' ').length;
                    timeToRead = Math.ceil(wordCount / wordsPerMinute)*60;
                } else {
                }
            },
        } );

        $.fn[ pluginName ] = function( options ) {
            return this.each( function() {
                if ( !$.data( this, "plugin_" + pluginName ) ) {
                    $.data( this, "plugin_" +
                        pluginName, new Plugin( this, options ) );
                }
            } );
        };

} )( jQuery, window, document );


jquery.kpi.js
;( function( $, window, document, undefined ) {
    "use strict";
        // Create the defaults once
        var pluginName = "kpi",
            defaults = {
                engagement: false,
                mouseLeave: false,
                pageCounter: 0,
                timeToRead: 0,
                timeToTrigger: 0,
                fadeDuration: 250,
                fadeDelay: 0,
                pageHits: 0,
                timeToRead: 0,
                regex: ''
            };

        function Plugin ( element, options ) {
            this.element = element;
            this.settings = $.extend( {}, defaults, options );
            this._defaults = defaults;
            this._name = pluginName;
            this.init();
        }

        if (sessionStorage.getItem("popped") === undefined) {
            var m_trigger = 0;
            sessionStorage.setItem("popped", m_trigger);
        } else {
            var m_trigger = sessionStorage.getItem("popped");
        }

        if (mCounter === undefined){
            var mCounter = 0;
        }

        if (timeToReadCounter === undefined){
            var timeToReadCounter = 0;
        }

        if (sessionStorage.getItem("pages") === undefined) {
            sessionStorage.setItem("pages", 1);
        }

        if (sessionStorage.getItem("newsletter") === undefined) {
            sessionStorage.setItem("newsletter", 1);
        }

        $.extend( Plugin.prototype, {
            init: function() {

                if (this.detectUrlRegex()){
                    return;
                }

                this.modalCounter();
                this.counter();
                this.subscribed();
                this.countPageHits();

                this.isMouseLeave();
                if (this.settings.engagement === true){
                    this.checkWordCount();
                 }
        },
            timeOut: function()
            {
                var id = this.element.id;

                if (this.settings.engagement === false){
                    window.setTimeout(function(){
                        $('#'+id).modal({
                            fadeDuration: this.settings.fadeDuration,
                            fadeDelay: this.settings.fadeDelay
                        });
                    }, 2000);
                }
                m_trigger++;
                sessionStorage.setItem("popped", m_trigger);
            },

            isMouseLeave: function()
            {
                var id = this.element.id;
                var trigger = sessionStorage.getItem("popped");

                if (this.settings.mouseLeave === false ){
                    return;
                }

                $('body').mouseleave(function(){
                    var activated = $('#coupon-cards');
                    if (this.settings.engagement === false && trigger <= 2 && mCounter <= 3) {
                        if ( activated.length && activated.css('display') ==  'none') {
                            $('#'+id).modal({
                                fadeDuration: this.settings.fadeDuration,
                                fadeDelay: this.settings.fadeDelay
                            });
                            setCookies('modalpopupcookie', 3);
                            setCookies('modalpageopened', 3);

                            m_trigger++;
                            sessionStorage.setItem("popped", m_trigger);
                        }
                    }
                });

                var setCookies = function(cookie, expiresIn) {
                    if (Cookies.get(cookie) === undefined) {
                        var cval = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
                        Cookies.set(cookie, cval, { expires: expiresIn, path: '/' });
                    }
                }
            },
            counter: function ()
            {
                var id = this.element.id;
                var timeToRead = this.settings.timeToRead;

                if (timeToRead <= 0){
                    return;
                }

                var runCounter = function(){
                    if (timeToReadCounter == timeToRead) {
                        clearInterval(this);
                    }
                    else {
                        timeToReadCounter++;
                        console.log("Time to read counter " + timeToReadCounter);
                        checkConditions();
                    }
                };
                // engagement is true
                var checkConditions = function () {
                    if (timeToReadCounter >= 7 && isCookieSet !== true){
                        $('#'+id).modal({
                            fadeDuration: fadeDuration,
                            fadeDelay: fadeDelay,
                        });
                    }
                    else if(timeToReadCounter*7 < Math.trunc(timeToRead*(2/3)) ){
                        ('#'+id).modal({
                            fadeDuration: fadeDuration,
                            fadeDelay: fadeDelay,

                        });
                    }
                };
                setInterval(runCounter, 6000);
                runCounter();
            },

            modalCounter: function(){
                var pageCounter = this.settings.pageCounter;
                if (pageCounter <= 0 ){
                       return;
                }

                var runModalCounter = function(){
                    if (mCounter == pageCounter) {
                        return;
                    } else {
                        mCounter++;
                        console.log(mCounter);
                    }
                };

                setInterval(runModalCounter, 1000);
                runModalCounter();
            },

            // get word count using p nodes
            checkWordCount: function()
            {
                var timeToRead = this.settings.timeToRead;
                var wordsPerMinute = 250;
                var nodes = document.querySelectorAll('p'),i;
                var gross = "";
                var wordCount = null;
                if(nodes.length > 0){
                    for (i = 0; i < nodes.length; i++) {
                        gross += nodes[i].innerText;
                    }
                    wordCount = gross.replace(/(<([^>]+)>)/ig,"").split(' ').length;
                    timeToRead = Math.ceil(wordCount / wordsPerMinute)*60;
                }
            },
            countPageHits: function(){

                var pageHits = this.settings.pageHits;

                if (sessionStorage.getItem("pages") !== null){
                      var pageHitsCurrent = sessionStorage.getItem("pages");
                } else {
                    var pageHitsCurrent = 1;
                }

                var id = this.element.id;

                 $(window).on('unload', function() {
                    pageHitsCurrent++;
                    sessionStorage.setItem("pages", pageHitsCurrent);
                });

                if (pageHits == 0 || pageHitsCurrent > pageHits ) {
                    return;
                }

                if (pageHits == pageHitsCurrent && Cookies.get('newsletter') === undefined ){
                      $('#'+id).modal({
                         fadeDuration: this.settings.fadeDuration,
                         fadeDelay:  this.settings.fadeDelay,
                    });
                }
            },
            subscribed: function() {
                 $('input[type="submit"]').on('click', function() {
                    if ($(this).val().toLowerCase() == 'subscribe'){
                        setCookies('newsletter', 365);
                        sessionStorage.setItem("newsletter",  1);
                    }
                });

                var setCookies = function(cookie, expiresIn) {
                    if (Cookies.get(cookie) === undefined) {
                        var cval = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
                        Cookies.set(cookie, cval, { expires: expiresIn, path: '/' });
                    }
                }
            },
            detectUrlRegex: function () {
                if (undefined === this.settings.regex || !this.settings.regex.length){
                    return;
                }
                var re = new RegExp(this.settings.regex);
                var sstring = window.location.pathname;
                var clean = sstring.replace(/\//g, '');

                if (clean.match(re) !== null){
                    return true;
                }

                return false;
            },
        });

        $.fn[ pluginName ] = function( options ) {
            return this.each( function() {
                if ( !$.data( this, "plugin_" + pluginName ) ) {
                    $.data( this, "plugin_" +
                       pluginName, new Plugin( this, options ) );
                }
            } );
        };

} )( jQuery, window, document );
